File: content/LLD/Reference-Code/15. Prototype Pattern/good_example.py
from typing import List
import copy


class ChessPiece:
    def __init__(self, name: str, position: str, color: str):
        self.name = name
        self.position = position
        self.color = color

    def display(self):
        return f"{self.color} {self.name} is at {self.position}"

    def clone(self):
        return copy.deepcopy(self)
        # return ChessPiece(self.name, self.position, self.color)


class ChessBoard:
    def __init__(self):
        self.pieces: List[ChessPiece] = []

    def add_piece(self, piece: "ChessPiece"):
        self.pieces.append(piece)

    def display_board(self):
        print("\nBoard State")
        for piece in self.pieces:
            print(f"   {piece.display()}")

    def clone(self):
        return copy.deepcopy(self)


piece1 = ChessPiece("King", "e2", "Black")
piece2 = ChessPiece("Queen", "a1", "Black")
piece3 = ChessPiece("King", "c3", "White")
piece4 = ChessPiece("Queen", "d2", "White")

chess_board = ChessBoard()
chess_board.add_piece(piece1)
chess_board.add_piece(piece2)
chess_board.add_piece(piece3)
chess_board.add_piece(piece4)

chess_board.display_board()


new_chess_board = chess_board.clone()


print("------------")
new_chess_board.add_piece(ChessPiece("Soldier", "d7", "black"))
new_chess_board.display_board()
chess_board.display_board()

File: content/LLD/Reference-Code/15. Prototype Pattern/GoodExample.java
import java.util.ArrayList;
import java.util.List;

public class GoodExample {
    interface Prototype<T> {
        T clone();
    }

    static class ChessPiece implements Prototype<ChessPiece> {
        private String name;
        private String position;
        private String color;

        public ChessPiece(String name, String position, String color) {
            this.name = name;
            this.position = position;
            this.color = color;
        }

        public String display() {
            return this.color + " " + this.name + " is at " + this.position;
        }

        @Override
        public ChessPiece clone() {
            return new ChessPiece(this.name, this.position, this.color);
        }
    }

    static class ChessBoard implements Prototype<ChessBoard> {
        public List<ChessPiece> pieces = new ArrayList<>();

        public void addPiece(ChessPiece piece) {
            this.pieces.add(piece);
        }

        public void displayBoard() {
            System.out.println("\nBoard State");
            for (ChessPiece piece : this.pieces) {
                System.out.println("   " + piece.display());
            }
        }

        @Override
        public ChessBoard clone() {
            ChessBoard copy = new ChessBoard();
            for (ChessPiece piece : this.pieces) {
                copy.addPiece(piece.clone());
            }
            return copy;
        }
    }

    public static void main(String[] args) {
        ChessPiece piece1 = new ChessPiece("King", "e2", "Black");
        ChessPiece piece2 = new ChessPiece("Queen", "a1", "Black");
        ChessPiece piece3 = new ChessPiece("King", "c3", "White");
        ChessPiece piece4 = new ChessPiece("Queen", "d2", "White");

        ChessBoard chessBoard = new ChessBoard();
        chessBoard.addPiece(piece1);
        chessBoard.addPiece(piece2);
        chessBoard.addPiece(piece3);
        chessBoard.addPiece(piece4);

        chessBoard.displayBoard();

        ChessBoard newChessBoard = chessBoard.clone();

        System.out.println("------------");
        newChessBoard.addPiece(new ChessPiece("Soldier", "d7", "black"));
        newChessBoard.displayBoard();
        chessBoard.displayBoard();
    }
}
